### PR TITLE
lazy-load nvrtc and libcuda

### PR DESCRIPTION
**Pull request by soumith**
Fixes https://github.com/pytorch/pytorch/issues/3395

### CODE REVIEW

**Comment by ezyang** [comment]
Nice! I don't exactly understand the mechanics; but I am wondering, is `TorchNvrtcMethods` intended to always be empty?

**Comment by soumith** [comment]
yea that's just a placeholder really.

File: torch/__init__.py [hunk 1]
```diff
@@ -48,8 +48,11 @@
         import torch._dl as _dl_flags
 
 old_flags = sys.getdlopenflags()
-sys.setdlopenflags(_dl_flags.RTLD_GLOBAL | _dl_flags.RTLD_NOW)
-
+sys.setdlopenflags(_dl_flags.RTLD_GLOBAL | _dl_flags.RTLD_LAZY)
+try:
+    import torch._nvrtc
```
Comment (apaszke) [inline torch/__init__.py]: What's the difference between this and just always linking nvrtc into `_C` and opening it with `RTLD_LAZY`? Wouldn't that work too? It seems a lot simpler

File: setup.py [hunk 2]
```diff
@@ -592,6 +592,16 @@ def make_relative_rpath(path):
...
 if WITH_CUDA:
+    THNVRTC = Extension("torch._nvrtc",
+                        libraries=['nvrtc', 'cuda'],
+                        sources=['torch/csrc/nvrtc.cpp'],
+                        language='c++',
+                        include_dirs=include_dirs,
+                        library_dirs=library_dirs + [cuda_lib_path + '/stubs'],
+                        extra_link_args=extra_link_args + [make_relative_rpath('lib')],
+                        )
+    extensions.append(THNVRTC)
```
Comment (apaszke) [inline setup.py]: you might need to build with `-Wl,--no-as-needed`. You're not using any of `libnvrtc` or `libcuda` symbols

File: torch/__init__.py (hunk 1 above)
Comment (soumith) [reply torch/__init__.py]: that does not work. it still tries to open libnvrtc.so immediately when it loads _C.so. (I tried it first.)

File: setup.py (hunk 2 above)
Comment (soumith) [reply setup.py]: fixed on master

**Pull request merged**